src/context.js

```diff
@@ -0,0 +1,191 @@
+import { ControlFunction, HaltError } from './control';
+
+export class ExecutionContext {
+  static ids = 1;
+  get isUnstarted() { return this.state === 'unstarted'; }
+  get isRunning() { return this.state === 'running'; }
+  get isWaiting() { return this.state === 'waiting'; }
+  get isCompleted() { return this.state === 'completed'; }
+  get isErrored() { return this.state === 'errored'; }
+  get isHalted() { return this.state === 'halted'; }
+
+  get isBlocking() { return this.isRunning || this.isWaiting || this.isUnstarted; }
+
+  get hasBlockingChildren() {
+    for (let child of this.children) {
+      if (child.isBlocking) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  constructor(parent = undefined, options = {}) {
+    this.id = this.constructor.ids++;
+    this.parent = parent;
+    this.options = options;
+    this.children = new Set();
+    this.exitHooks = new Set();
+    this.state = 'unstarted';
+    this.resume = this.resume.bind(this);
+    this.fail = this.fail.bind(this);
+    this.ensure = this.ensure.bind(this);
+    this.call = this.call.bind(this);
+  }
+
+  get promise() {
+    this._promise = new Promise((resolve, reject) => {
+      this.resolve = resolve;
+      this.reject = reject;
+    });
+    this.finalizePromise();
+    return this._promise;
+  }
+
+  finalizePromise() {
+    if(this.isCompleted && this.resolve) {
+      this.resolve(this.result);
+    } else if(this.isErrored && this.reject) {
+      this.reject(this.result);
+    } else if(this.isHalted && this.reject) {
+      this.reject(new HaltError(this.result));
+    }
+  }
+
+  then(...args) {
+    return this.promise.then(...args);
+  }
+
+  catch(...args) {
+    return this.promise.catch(...args);
+  }
+
+  finally(...args) {
+    return this.promise.finally(...args);
+  }
+
+  get root() {
+    if (!this.parent) {
+      return this;
+    } else {
+      return this.parent.root;
+    }
+  }
+
+  call(operation, options = {}) {
+    let child = new ExecutionContext(this, options);
+    this.children.add(child);
+    child.ensure(() => {
+      this.children.delete(child);
+      let { result } = child;
+      if (this.isRunning || this.isWaiting) {
+        if (child.isErrored) {
+          if (typeof options.fail === 'function') {
+            options.fail(result, this.fail);
+          } else {
+            this.fail(result);
+          }
+        } else if (child.isHalted) {
+          if (typeof options.halt === 'function') {
+            options.halt(result);
+          } else {
+            this.fail(new HaltError(result));
+          }
+        }
+        if (this.isWaiting && !this.hasBlockingChildren) {
+          this.finalize('completed');
+        }
+      }
+    });
+    child.enter(operation);
+    return child;
+  }
+
+  ensure(hook) {
+    let { exitHooks } = this;
+    exitHooks.add(hook);
+    return () => exitHooks.delete(hook);
```

Comment:
If we ever re-assign `this.exitHooks`, then holding onto this reference could actually be a bit problematic, I think? Maybe it's better to just use `this.exitHooks` directly?